src/lib/Embed.hs

```diff
@@ -736,6 +781,37 @@ intToIndexE ty@(TC con) i = case con of
     return $ Con $ SumCon isLeft li ri
   _ -> error $ "Unexpected type " ++ pprint con
   where iAsIdx = return $ Con $ Coerce ty i
+intToIndexE (RecordTy types) i = do
+  sizes <- traverse indexSetSizeE types
+  (strides, _) <- traverseFunM
+    (\sz prev -> do {v <- imul sz prev; return ((prev, v), v)}) sizes (IntVal 1)
+  offsets <- flip mapM (zip (toList types) (toList strides)) $
+    \(ty, (s1, s2)) -> do
+      x <- irem i s2
+      y <- idiv x s1
+      intToIndexE ty y
+  return $ Record (restructure offsets types)
+intToIndexE rty@(VariantTy types) i = do
+  sizes <- traverse indexSetSizeE types
+  (offsets, _) <- traverseFunM
+    (\sz prev -> do {v <- iadd sz prev; return (prev, v)}) sizes (IntVal 0)
+  let
+    reflect = reflectLabels types
+    -- Find the right index by looping through the possible offsets
+    -- TODO: is there a better way than generating a case statement? Select
+    -- doesn't work because variants aren't primitive types
+    go prev ((label, repeatNum), ty, offset) = do
+      beforeThis <- unsafeCoerce PreludeBoolTy =<< ilt i offset
+      hereCase <- buildNAbs Empty $ \[] -> do
+        shifted <- isub i offset
+        index <- intToIndexE ty shifted
+        return $ Variant types label repeatNum index
+      beforeCase <- buildNAbs Empty $ \[] -> return $ prev
+      emit $ Case beforeThis [hereCase, beforeCase] rty
```

Comment:
This looks like a good candidate for a standalone function like

```hs
boolCase :: MonadEmbed m => Atom -> m Atom -> m Atom -> m Atom
boolCase intCond trueBlock falseBlock = ...
```

Reply:
Also, IMO the case to `PreludeBoolTy` is ok!

Reply:
I guess you still need to pass in the return type, right? Since that gets passed to `Case`. (Or maybe there's a way to easily infer that from the `trueBlock` and `falseBlock`? I guess we could wrap the case in a `Lam` and then check its type, but that seems a bit annoying...)

Reply:
hm, actually, maybe I can just `getType body` for the body of the `Abs`?

Reply:
Yep, you can just query the type. (`getType` doesn't need to know about the variables in scope since their types are cached at the occurrences.)